File: lvdou-user-web/src/main/java/com/lvdou/controller/UserController.java
package com.lvdou.controller;

import com.lvdou.pojo.Address;
import com.lvdou.pojo.User;
import com.lvdou.service.impl.UserServiceImpl;
import org.apache.commons.lang3.StringUtils;
import org.apache.shiro.SecurityUtils;
import org.apache.shiro.authc.AuthenticationToken;
import org.apache.shiro.authc.IncorrectCredentialsException;
import org.apache.shiro.authc.UnknownAccountException;
import org.apache.shiro.authc.UsernamePasswordToken;
import org.apache.shiro.subject.Subject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/user")
public class UserController {

    @Autowired
    private  UserServiceImpl userService;

    /** 注册用户 */
    @PostMapping("/saveUser")
    public Map<String,Object> save(@RequestBody User user,
                                String address, String vcode){

       return userService.saveUserAndAddress(user,address,vcode);


    }

    /** 发送短信验证码 */
    @GetMapping("/sendCode")
    public boolean sendCode(String phone){
        try {
            if(StringUtils.isNoneBlank(phone)&&phone.length()>0){
                /** 发送验证码 */
                userService.sendValidate(phone);
            }
            return true;
        }catch (Exception ex){
            System.out.println("调用短信验证码失败！");
            ex.printStackTrace();
        }
        return false;
    }
    @PostMapping("/login")
    public Map<String,Object> login(@RequestBody User user,
                        @RequestParam("vcode") String vcode){
         String phone = user.getPhone();
        Map<String, Object> map = new HashMap<>();
        //String mobileReg="^1[3|5|7|8]\\d{9}$";
             if(phone!=null&&phone.length()==11){
                 Subject subject = SecurityUtils.getSubject();

                 //2.执行认证方法
                 AuthenticationToken token = new UsernamePasswordToken(phone,
                         "");
                 try {
                     subject.login(token);

                     //把用户数据存入session（不是给shiro作为登录标记的，而是用于我们自己的业务逻辑）
                     User loginUser = (User)subject.getPrincipal();

                     int a =userService.saveUserToRedis(loginUser);
                     if(a==0){
                         System.out.println("用户存入redis成功");
                     }
                     map.put("user",user);
                     return map;
                 } catch (UnknownAccountException e) {
                     map.put("msg","账户不存在！");
                 } catch (IncorrectCredentialsException e) {
                     System.out.println("密码错误！！！！！wss");
                    map.put("msg","密码错误");
                 } catch (Exception e) {
                     System.out.println("系统错误");
                     e.printStackTrace();
                    map.put("msg","系统异常");
                 }
                 //1:代表登录成功

             }
             map.put("msg","手机号码格式有误！");
             return map;

    }
    @PostMapping("/login2")
    public Map<String,Object> login2(@RequestBody User user
                        ,@RequestParam("vcode") String vcode){
            return  userService.checkVCode(user, vcode);
    }

    @GetMapping("/checkUserName")
    public Map<String,Object> checkUserName(String userName){
        return  userService.checkUserName(userName);
    }
    @PostMapping("/registerUser")
    public Map<String,String> registerUser(User user){
        return userService.registerUser(user);
    }


}

File: lvdou-user-web/src/main/webapp/js/controller/userController.js
/** 定义控制器层 */
app.controller('userController', function($scope, baseService){

    // 发送短信验证码
    $scope.sendCode = function(){
        // 判断手机号码
        if ($scope.user.phone && /^1[3|5|7|8]\d{9}$/.test($scope.user.phone)){
            // 发送异步请求
            baseService.sendGet("/user/sendCode?phone=" + $scope.user.phone)
                .then(function(response){
                    if (response.data==true){
                        alert("发送成功！请注意查收");
                    }
                },function () {
                    alert('发送失败！');
                });
        }else{
            alert("手机号码格式不正确！");
        }
    };
    //登录
    $scope.login=function(){
        baseService.sendPost("/user/login2?vcode="+$scope.vcode,$scope.user)
            .then(function (response) {
                if(response.data.user){
                    $scope.user=response.data.user;
                    location.href = "/index.html";
                    alert("登录成功！");
                }else {
                    alert(response.data.msg);
                }
            },function () {
                alert('亲，作为日理万鸡的系统，我很忙啦，请稍后再试！');
            })
    };


    $scope.checkUserName=function () {
        baseService.sendGet("/user/checkUserName?userName="+$scope.user.username)
            .then(function (response) {
                if(response.data.checkResult=="1"){
                    alert("用户名可以使用！")
                }else {
                    alert("用户名不可以使用！")
                }
            },function () {
                alert("出现异常，请重试！")
            })
    };
    $scope.saveUser=function () {
        if($scope.address2==null||$scope.address2==""){
            baseService.sendPost("/user/saveUser?vcode="+$scope.vcode+"&address="+
                $scope.address1,$scope.user)
                .then(function (response) {
                    if(response.data.user){

                        alert("注册成功！");
                        location.href = "/index.html";
                    }else {
                        alert(response.data.msg);
                    }
                },function () {
                    alert("连接异常，请重试！");
                });
        }else{
            alert('请先不要输入第二个地址！')
        }

    }
});